[KengHian] iP

KengHian opened the pull request:
### StrugglingPro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

StrugglingPro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

KengHian committed:
chore: rename Duke to Struggling

KengHian committed:
feat: greeting and exit

KengHian committed:
feat: echo command

KengHian committed:
feat: add and list input

KengHian committed:
feat: add Task class

KengHian committed:
feat: add, mark and unmark tasks

KengHian committed:
feat: add Deadline, Event and ToDo classes

KengHian committed:
chore: rename taskDescription to description

KengHian committed:
feat: support tracking of tasks

KengHian committed:
test: add text ui testing

KengHian committed:
feat: add minimal error handling

KengHian committed:
feat: add capability to delete task

KengHian committed:
refactor: use enums for acceptable commands

KengHian committed:
test: add exception and delete cases

KengHian committed:
chore: ignore savefile

KengHian committed:
feat: Add functions for all Task to report state

KengHian committed:
feat: Add ability for chatbot to save Tasks in a file

KengHian committed:
Merge branch 'branch-Level-7'

KengHian committed:
feat: Store Deadline by with LocalDate

KengHian committed:
Merge branch 'branch-Level-8'

KengHian committed:
feat: Add classes for abstraction

KengHian committed:
Merge branch 'branch-A-MoreOOP'

KengHian committed:
chore: Update test cases

KengHian committed:
feat: Put all classes into their respective packages

KengHian committed:
Merge branch 'add-gradle-support'

KengHian committed:
feat: Convert option to array index in Parser for clearer responsibility

KengHian committed:
test: Add unit test for Parser and TaskList class

KengHian committed:
Merge branch 'branch-A-JUnit'

KengHian committed:
chore: Remove unused code in Struggling class

KengHian committed:
chore: Setup gradle build config

KengHian committed:
docs: Add Javadoc for public classes and methods

KengHian committed:
style: Implement code style

KengHian committed:
feat: Add command to find task

KengHian committed:
Merge branch 'branch-A-JavaDoc'

KengHian committed:
Merge branch 'branch-A-CodingStandard'

KengHian committed:
case find:

KengHian committed:
style: Fix indentation

nihalramesh commented on src/main/java/struggling/Parser.java:
```diff
@@ -0,0 +1,86 @@
+            return new FindCommand(args[1]);
+        }
+
+        return new InvalidCommand();
```
Perhaps instead of placing this line outside of the code block, place it within a default case ?

nihalramesh commented on src/main/java/struggling/Parser.java:
```diff
@@ -0,0 +1,86 @@
+ * Commands accepted by the Parser class.
+ */
+enum Commands {
+    bye, list, mark, unmark, todo, deadline, event, delete, find
```
enum names should ideally be in PascalCase

nihalramesh commented on src/main/java/struggling/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+     * @return The Task removed.
+     */
+    public Task deleteTask(int i) {
+        Task t = this.tasks.remove(i);
```
Perhaps we should check whether the provided index i is valid before removing ?

nihalramesh commented on src/main/java/struggling/Ui.java:
```diff
@@ -0,0 +1,124 @@
+     * Displays greeting when the program terminates.
+     */
+    public void showGoodBye() {
+        show("Bye. Hope to see you again soon!");
```
Perhaps we should close the scanner after usage

nihalramesh reviewed (commented):
Overall, your code was quite easy to read, just a bit of nitpicking. Other than that, was mostly good!

linette-g commented on src/main/java/struggling/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+     * representation of each Task in TaskList.
+     */
+    public ArrayList<String> getTasksString() {
+        ArrayList<String> arr = new ArrayList<>();
```
Perhaps could use a more descriptive name for this variable?

linette-g commented on src/main/java/struggling/Struggling.java:
```diff
@@ -0,0 +1,66 @@
+                Command c = Parser.parse(cmd);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
+            } catch (IllegalArgumentException e) {
+                ui.showError("OOPS!!! I'm sorry, but I don't know what that means :-(");
+            } catch (StrugglingException e) {
+                ui.showError(e.getMessage());
+            } catch (IOException e) {
+                ui.showError("Fail to save task, please contact developer!");
+            }
```
I like the error handling here!

linette-g commented on src/main/java/struggling/command/ExitCommand.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    @Override
+    public boolean isExit() {
```
I like the clear naming of boolean methods 👍

linette-g commented on src/main/java/struggling/Ui.java:
```diff
@@ -0,0 +1,124 @@
+        show(ans.toString());
+    }
+
+    public void find(ArrayList<String> arr) {
```
Perhaps could add a header comment for this method?

linette-g reviewed (commented):
LGTM! Overall code is quite comprehensive.

KengHian committed:
style: Add checkstyle file and enforce the standard

KengHian committed:
Merge branch 'branch-A-CheckStyle'

KengHian committed:
feat: Add GUI for chatbot

KengHian committed:
Merge branch 'branch-Level-10'

KengHian committed:
chore: Add assert after internal functions

KengHian committed:
refactor: Remove dead code in Struggling class

There are several blocks of dead code in Struggling class
as the project adapt to the GUI interation.

Removing  all those dead code improves the code quality.

Let's remove the run() and main() method in Struggling
class as they are no longer needed.

KengHian committed:
refactor: Extract arguments processing for tasks

In Parser class there are blocks of code in switch
block processing user input when before creating
the various tasks.

Extracting methods helps improve code quality and readability.

Let's extract extact them into their respective
methods and call them from the switch block.

KengHian committed:
refactor: Replace magic literal in Parser class

There are a few specific numbers, which are the index
of specific arguments in user command, used directly
without context.

Magic literals might cause confusion and replacing
them would improve the code quality.

Let's create a few constant variales for index with
fixed values and intermidiate variables for index
with dynamic values to provide necessary contexts.

KengHian committed:
feat: Use stream in TaskList to map and filter

KengHian committed:
Merge pull request #1 from KengHian/branch-A-Assertions

chore: Add assert after internal functions

KengHian committed:
Merge branch 'master' into branch-A-CodeQuality

KengHian committed:
Merge pull request #2 from KengHian/branch-A-CodeQuality

Branch a code quality

KengHian committed:
Merge pull request #3 from KengHian/branch-A-Streams

feat: Use stream in TaskList to map and filter

KengHian committed:
chore: Setup Github workflow

KengHian committed:
chore: Add code for ui testing

KengHian committed:
fix: Fix index bug when parsing cmd

KengHian committed:
fix: Update Struggling main location for text ui test

KengHian committed:
feat: Enable setting of task priority

KengHian committed:
Merge branch 'branch-C-Priority'

KengHian committed:
chore: Disable mac os text ui test case

KengHian committed:
feat: Improve UI style

KengHian committed:
feat: Changes response phrases for personality

KengHian committed:
refactor: Remove dead code

KengHian committed:
refactor: Fix header comments

KengHian committed:
chore: Set up a product website

KengHian committed:
chore: Add section link

KengHian committed:
feat: Reduce DialogBox size for compactness

KengHian committed:
feat: Auto save on change

KengHian committed:
bug: Fix user guide seciton link

KengHian committed:
test: Add tests for task priority functions

KengHian committed:
docs: Add more details

KengHian committed:
chore: Fix user guide

KengHian committed:
feat: Save data on exit

KengHian committed:
docs: Fix user guide
